### PR TITLE
Call subcommands from zsh completion

### PR DESCRIPTION
**Pull request by joshka**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [ ] Have you successfully run `brew tests` with your changes locally?

(I'm not sure that these changes require tests so much as a code review and
a look at the other related changes.)

-----

This change is inspired by the way that the git zsh completions work by foisting
the responsibility for sub commands onto the command themselves.
It is paired with another change that takes the oh-my-zsh brew cask plugin and
makes it into a first class completion rather than it overriding this.

This change requires fixes in brew, oh-my-zsh and the zsh-completions, and will
likely cause the `brew cask` completions to break until these have been fully
accepted into each project. I'd anticipate doing brew first and the others after.

See related PRs:
https://github.com/zsh-users/zsh-completions/pull/430
https://github.com/robbyrussell/oh-my-zsh/pull/5191

### CODE REVIEW

**Comment by joshka** [comment]
Anyone care to comment on this?

File: share/zsh/site-functions/_brew [hunk 1]
```diff
@@ -29,6 +27,12 @@ _brew_outdated_formulae() {
   outdated_formulae=(`brew outdated`)
 }
 
+__brew_command() {
+  local command=$1
```
Comment (UniqMartin) [inline share/zsh/site-functions/_brew]: I picked this line, but please always quote uses of a variable in the parts you modified/extended. We prefer to do that for consistency and in this case also for correctness because we're dealing with user input. In this particular example: Use `"$1"` instead of `$1`.

**Comment by joshka** [comment]
@UniqMartin fixed

**Comment by UniqMartin** [comment]
I'm sorry, but the feedback I will be able to provide is limited because I'm not a ZSH user. Do I understand correctly, that the gist of this change is: if none of the built-in command completions apply, check whether a shell function named `_brew_<command>` exists and if it does, let it handle the completion?

If that's the case, I think this is a reasonable change. Ideally, and if you don't consider this to be out of scope, I'd love to see this PR extended such that the completion script is able to automatically load a completion file for an external command provided by a tap (as this tends to be the primary way of distributing external commands for `brew` nowadays). A very rough sketch would be:

1. If applicable, use the built-in completion and be done.
2. If it exists, use the `_brew_<command>` shell function and be done.
3. If `brew command <command>` returns a path that points into the `cmd/` directory of a tap, use that to construct a path like `<tap>/cmd/completions/zsh/_brew_<command>` and try to source the file. Then retry step 2. (The exact path for hosting those completions in a tap is obviously up for debate.)

**Comment by UniqMartin** [comment]
> fixed

Nope, it's not. Please do this for all the variables in all the lines you touched (as I asked you in my comment), not just the single line I commented on. Thanks!

**Comment by joshka** [comment]
My mistake - I misread your comment. Sorry about that. Should be all fixed now :)

Yes, you understand correctly. I'm flipping the responsibility for calling the brew cask completions to be from the brew completion rather than the other way around. The general convention for completions in zsh is that _command() defines a completion, and these functions are generally loaded automatically by the shell by sticking them somewhere in fpath and running compinit.

On your suggestion for extending this, I **think** it's possibly out of scope for this change. Completions installed as part of a tap should be symlinked to /usr/local/share/zsh/site-functions and hence be automatically loaded by the shell. I haven't confirmed this however. Can you please point me to an example of a tap completion that I can look at?

**Comment by UniqMartin** [comment]
> My mistake - I misread your comment. Sorry about that. Should be all fixed now :)

It is. Thanks!

> The general convention for completions in zsh is that _command() defines a completion, and these functions are generally loaded automatically by the shell by sticking them somewhere in fpath and running compinit.

What is triggering the actual load? I imagine if one types `ls` for the first time then the logic behind `compinit` will make sure to load `_ls` from `fpath` to provide those completions. Am I right?

But how does the auto-loading work for something like `brew cask`? I imagine the load would need to be triggered from within `_brew`. Is doing `declare -f _brew_cask` sufficient to make ZSH look into `fpath` and load a file named `_brew_cask` if there's no function of that name yet? (Sorry for asking possibly stupid questions, but I figure this might still be faster than me becoming a ZSH user. :wink:)

> On your suggestion for extending this, I **think** it's possibly out of scope for this change.

That's too bad. I think this would be really useful and it would be awesome if someone made the first step for the benefit of all external commands hosted in taps.

> Completions installed as part of a tap should be symlinked to /usr/local/share/zsh/site-functions and hence be automatically loaded by the shell.

This partially taps into the above question regarding auto-loading. But regardless, I think symlinking is a bit more error prone than what I suggested and would require Homebrew to manage yet another set of symlinks in this directory, in addition to completions provided by Homebrew itself (just `_brew`) and completions provided by formulae. Obviously, that's just my opinion and other maintainers may strongly disagree. But to get back to being constructive, would automatically extending `fpath` from within `_brew` be a viable alternative to symlinking?

> Can you please point me to an example of a tap completion that I can look at?

Unfortunately, I cannot. I'm not aware of any taps with commands that define their own completions in this way, simply because the Homebrew completion script doesn't offer this level of extensibility (yet).

**Comment by joshka** [comment]
> What is triggering the actual load? I imagine if one types ls for the first time then the logic behind compinit will make sure to load _ls from fpath to provide those completions. Am I right?

Kinda - compinit sets up the autoload, but zsh manages the autoload when you do this.

> But how does the auto-loading work for something like brew cask? I imagine the load would need to be triggered from within _brew. Is doing declare -f _brew_cask sufficient to make ZSH look into fpath and load a file named _brew_cask if there's no function of that name yet? (Sorry for asking possibly stupid questions, but I figure this might still be faster than me becoming a ZSH user. 😉)

compinit runs and looks at the first couple of lines of each of the scripts in fpath. It checks each for autoload. If it's an autoloaded function it interacts with the default zsh autoload feature (load script from fpath when necessary). The `declare -f` line allows us to only call the autoloaded function if it's been previously defined / setup for autoload. If the function doesn't exist, it returns non-zero, so this does nothing if there's no other command defined. Otherwise, calling the function causes autoload to define the function body properly (and read the _brew_cask script / function in full) See https://github.com/joshka/zsh-completions/commit/31b4e82887626eee236a9806ec9c4f5dee90056c for the brew cask side of the funciton.

> That's too bad. I think this would be really useful and it would be awesome if someone made the first step for the benefit of all external commands hosted in taps.

I _personally_ think that the completions should belong in the homebrew cask repo rather than external, so that they can be versioned concurrently with the project, but the maintainers have suggested that they don't want it in the repo. See https://github.com/caskroom/homebrew-cask/issues/22434#issuecomment-229008519. @vitorgalvao might be able to provide more context on why this is so.

> This partially taps into the above question regarding auto-loading. But regardless, I think symlinking is a bit more error prone than what I suggested and would require Homebrew to manage yet another set of symlinks in this directory, in addition to completions provided by Homebrew itself (just _brew) and completions provided by formulae. Obviously, that's just my opinion and other maintainers may strongly disagree. But to get back to being constructive, would automatically extending fpath from within _brew be a viable alternative to symlinking?

My /usr/local/share/zsh/site-functions already contains a bunch of symlinks to installed formulae.
```
$ ll /usr/local/share/zsh/site-functions
total 72
-rw-r--r--  1 joshka  admin   5.2K Jun 29 17:28 _brew
-rw-r--r--  1 joshka  admin   3.1K Jun 27 16:17 _brew_cask
lrwxr-xr-x  1 joshka  admin    55B Jun 16 15:08 _git -> ../../../Cellar/git/2.9.0/share/zsh/site-functions/_git
lrwxr-xr-x  1 joshka  admin    60B Jun 26 17:49 _hg -> ../../../Cellar/mercurial/3.8.3/share/zsh/site-functions/_hg
lrwxr-xr-x  1 joshka  admin    55B Feb 11 10:32 _hub -> ../../../Cellar/hub/2.2.3/share/zsh/site-functions/_hub
lrwxr-xr-x  1 joshka  admin    56B Jun 26 17:49 _mpv -> ../../../Cellar/mpv/0.18.0/share/zsh/site-functions/_mpv
lrwxr-xr-x  1 joshka  admin    74B Jun 26 17:49 _youtube-dl -> ../../../Cellar/youtube-dl/2016.06.26/share/zsh/site-functions/_youtube-dl
lrwxr-xr-x  1 joshka  admin    70B Jun 16 15:08 git-completion.bash -> ../../../Cellar/git/2.9.0/share/zsh/site-functions/git-completion.bash
```
I'm assuming that this is the standard practice for formulae. I don't know how the tap side works and couldn't find the specific code that does this automatically. Perhaps I've misunderstood that particular part of the question. I think you probably could extend fpath, but it might make this a little harder to diagnose than a simple obvious symlink. If you suggested that bin folder of every formula ended up in your $PATH, you'd agree that that would be excessive I think. I'm also guided again by the way the git completion works in this regard. Consistency here seems better than doing something different.

File: share/zsh/site-functions/_brew [hunk 2]
```diff
@@ -145,4 +150,6 @@ case "$words[1]" in
       _brew_outdated_formulae
       _wanted outdated_formulae expl 'outdated formulae' compadd -a outdated_formulae
     fi ;;
+  *)
+    __brew_command "$command"
```
Comment (UniqMartin) [inline share/zsh/site-functions/_brew]: Another minor thing for consistency: Could you please add a trailing `;;` to terminate this case like all the others? (I know it's not strictly necessary.)

**Comment by UniqMartin** [comment]
Thank you for the detailed response and for your patience! I think the change you're proposing makes sense and I find your arguments convincing. Automatic discovery of completions for external commands from taps would certainly be a topic worth investigating further, but doesn't need to happen in this PR.

I tested your changes by additionally downloading the `_brew_cask` file you linked to. Everything worked as I would have expected. I'm inclined to merge your changes, but would feel more comfortable if any of the ZSH-using maintainers could briefly have a look.

@apjanke Do you happen to have time to review this? I think the changes proposed here are fine.

> I don't know how the tap side works and couldn't find the specific code that does this automatically.

There's currently no automatic way for taps, but it could become part of the `tap`/`untap` procedure, that currently already automatically handles man pages [in `Tap#link_manpages`](https://github.com/Homebrew/brew/blob/0926f920dc48e297d569363acbb46d1058c6ba73/Library/Homebrew/tap.rb#L249-L270) and [in `Tap#unlink_manpages`](https://github.com/Homebrew/brew/blob/0926f920dc48e297d569363acbb46d1058c6ba73/Library/Homebrew/tap.rb#L288-L296). Something similar could be done for completions, if they were hosted in the tap repository.

**Comment by joshka** [comment]
Fixed. No problem @UniqMartin. Better to get it done right :)

I agree with your idea about extending the tap link manpages idea to handle completions, but I think we can possibly get the same / better effect by moving the brew cask changes I suggested in zsh-completions to https://github.com/homebrew/homebrew-completions. This align's with the [comment](https://github.com/caskroom/homebrew-cask/issues/22434#issuecomment-229008519) by @vitorgalvao that  "Completions belong in other projects".

**Comment by UniqMartin** [comment]
> Fixed. No problem @UniqMartin. Better to get it done right :)

Thanks! Let's sit on this for a few more days and if nobody complains (or comments), I'll merge this. Feel free to ping me if there's no activity here and I happen to forget to deal with this.

> I agree with your idea about extending the tap link manpages idea to handle completions, but I think we can possibly get the same / better effect by moving the brew cask changes I suggested in zsh-completions to

Just to be clear, I think both approaches make sense and it's worth supporting both use cases. Cask would already be served by the change in this PR, but other projects might want to stick the completions into their tap repositories and make them an integral part of their commands. To me it makes a lot of sense to have them in the same repository and to update them together with commands and options being added and removed, but I respect if someone feels differently about this.

That said, that's often not happening when we perform those changes in Homebrew. Bash completions are seeing a bit more love (though they are also not exactly in sync), but ZSH completions are even worse. If you feel passionate about the latter, further PRs to improve them are more than welcome!

**Comment by apjanke** [comment]
This approach of breaking the subcommand completions out into per-subcommand functions makes sense, because that allows them to be maintained along with the subcommands and not require changes to the main `_brew` completion logic. Especially with inverting how the `cask` completion currently works.

I think it's worth trying to figure out how those subcommand completion functions are going to be made available as part of this PR, though. That's an important part of the completion system design; having the hook in place to call them isn't that useful if we don't have the subcommand completion functions themselves available. Merging this as is would kind of be leaving it "some assembly required" for users or other projects to figure out, and then if we do end up adding a mechanism for it later, it might be incompatible with what they've worked out, or would at least require a transition.

I like having a mechanism for taps to provide the completion functions for subcommands they supply, like a zsh completions directory of autoloaded functions that gets picked up automatically.

> My /usr/local/share/zsh/site-functions already contains a bunch of symlinks to installed formulae.

Yep, that's standard practice. `<prefix>/share/zsh/site-functions` is just another well-known directory like `bin/` or `man/man5/`, and formulae that install files there under their keg get them symlinked during installation.

Symlinking tap-provided completion functions there as part of the tap/untap/update process could work. That would be nice because it's the same location that formula-provided completions go to, so Zsh users probably already have that in their `$fpath` and they would just work without additional user configuration. (Homebrew developers who were writing completions for new commands would have to symlink them manually or do a `brew tap --repair` or something.) It has the downside that there could be collisions between command names in different paths, and resolving them under a symlink regime would be a little knottier than with a runtime path based mechanism. And it's one more set of symlinks to manage.

>  I think you probably could extend fpath, but it might make this a little harder to diagnose than a simple obvious symlink.

Adding to `$fpath` alone would mostly be useful if it were done at shell startup time, before `compinit` was called, so `compinit` could do the work of locating the functions and marking them as autoloadable. And Homebrew doesn't have a mechanism for "configure this ZSH session for `brew` use", so there's no hook for this. I.e., we leave it up to users to modify their startup files to get `#{HOMEBREW_PREFIX}/share/zsh/site-functions` on `$fpath` in the first place. So I think you'd be stuck doing a pseudo-autoloading process like Martin laid out in an earlier comment. Could be harder to diagnose things. But it also restricts the scope of changes to just zsh-specific code, and may be less risky.

Neither of these approaches is sticking out as the obvious better way to do things. But I think the symlink approach has the edge: for packages that provide zsh completions, installing them to `share/zsh/site-functions` seems to be the conventional way to do it. And it's analogous to how we handle man pages provided by taps; they're also command accessories so it makes sense to handle them similarly.

**Comment by apjanke** [comment]
As far as interactions with other projects: I don't know that it makes sense to put the `_brew_cask` completion in `zsh-users/zsh-completions`. Especially not before the hook mechanism it depends on has been accepted in Homebrew. That's an independent, third-party project that aims to supply Zsh completions for projects which don't supply their own. But we *are* the Homebrew project, so if we're going to write zsh completion support, we actually could supply it via our own repos. That way we have commit control over it, and can synchronize changes in the commands with corresponding changes in their completion definitions.

I also think it generally works best to have completions supplied as part of the package that defines the command. They're like man pages: accessories to the command which are version-sensitive and should generally be installed and updated along with the commands they go with. The situation with the existence of separate `*sh-completions` projects seems to be because many packages don't have expertise or interest in providing shell completions for their commands, especially for the less popular `fish` and `zsh`. (And these third-party projects can provide coverage for older versions of commands, where it's too late to go back and retroactively ship completion definitions with prior releases.)

Synchronizing a change like this across multiple organizations could be a pain. It would be nice if we could implement it all internal to Homebrew first. If Homebrew Cask doesn't want zsh completion definitions in their tap, maybe it would make more sense to put them in a `Homebrew/brew-cask-zsh-completions` repo under Homebrew's control? There's an existing [Homebrew/completions tap](https://github.com/homebrew/homebrew-completions) which has formulae for similar per-package completions packages. Then we could add a `brew-cask-zsh-completion` formula there for it, and you'd have the whole thing in-organization and independently versioned. (So changes to `_brew_cask` completion wouldn't have to wait for upstream `zsh-completions` to make a new release for them to be deployed to users.) (`Homebrew/completions` already has a `brew-cask-completion` formula which has bash and fish support, but it's from another third party project, so we probably wouldn't want to just add zsh support to that existing project.)

I'd also hold off on making a PR to Oh-My-Zsh for this. Their development is slow moving, and a PR submitted there before it's actually ready for action is likely to get lost in the backlog. You might wait until we've agreed on an approach here and actually shipped it before suggesting a deprecation over there.

**Comment by apjanke** [comment]
Oh, see also: https://github.com/Homebrew/legacy-homebrew/issues/44910. The `zsh-completions` package doesn't install its completions to `share/zsh/site-functions`; it installs them to `share/zsh-completions`, which the user must also manually get on to `$fpath`. It would be nice if `brew`'s autocompletion support didn't require that extra step, or a dependency on another project.

Also for background, https://github.com/Homebrew/legacy-homebrew/issues/49066, from when `zsh-completions` removed the previously-existing `_brew` completion and we had some breakage.

**Comment by joshka** [comment]
I think you're correct on the best approach for this. It makes sense to me to stick the `brew cask` completions in `Homebrew/brew-cask-zsh-completions` then reference from `homebrew/homebrew-completions` rather than the PRs I've put for zsh-completions / omz. I'll get on it.

**Comment by MikeMcQuaid** [comment]
> I'll get on it.

Any news?

**Comment by joshka** [comment]
I'm currently travelling for my birthday. I'll take a look in a couple of days from now

**Comment by joshka** [comment]
@MikeMcQuaid / @apjanke I've just pushed https://github.com/joshka/brew-cask-zsh-completions. If you have some time to take a quick look for me I'd really appreciate it. `homebrew/homebrew-completion` update is in the works.

**Comment by MikeMcQuaid** [comment]
@joshka Not sure what I'm looking at/for personally so will pass to @apjanke.

**Comment by joshka** [comment]
@pstadler / @KevinBongart can either of you can comment on this change, given your authorship of the oh-my-zsh plugin. For a little TL;DR context, I'm suggesting that rather than overriding the `_brew()` completion function from the `_brew_cask()` we flip this so that `_brew()` calls `_brew_cask()` (or any other brew subcommand).

I've also rewritten and cleaned up the `_brew_cask` completion and moved it to its own repo [joshka/brew-cask-zsh-completions](https://github.com/joshka/brew-cask-zsh-completions/blob/master/_brew_cask) which will be installable via the homebrew/homebrew-completions tap. This allows users that choose [prezto](https://github.com/sorin-ionescu/prezto) over oh-my-zsh to use the completions.

One thing I'd draw your attention to is I don't know if this break oh-my-zsh users usage.

@apjanke do you have any feedback on this?

**Comment by pstadler** [comment]
@joshka The current implementation is definitely hacky, because I didn't want to touch the actual brew plugin back then. Honestly, I'm not using cask on a frequent basis, you probably know better how close those two projects are these days and how much sense it makes to merge or refactor this.

**Comment by apjanke** [comment]
Haven't had time to read through the full thing, but I like this approach, and what I've read so far looks good. Left a comment on how to make a `test` for it.

Minor doco thing: The joshka/brew-cask-zsh-completions repo should have its own standalone installation instructions which are independent of OMZ/Prezto/brew, even if that's as minimal as "copy this to your Zsh `site-functions` directory", and should reference the `homebrew/completions` tap and installation method in its README.

I think once this is working well, we should consider folding your completion definitions in to `homebrew/brew` itself, now that we are integrating Cask in to core Homebrew. It was Cask's policy to keep `zsh` completions as a separate project, but not core Homebrew's policy: core supplies them as part of the main Homebrew project. So if we migrate Cask to Homebrew core, it would be nice to have the completions for it available by default (and versioned to match the commands they go with).

I will try to find time to read through and test the full code later today or tomorrow morning; I'm traveling at the moment.

**Comment by apjanke** [comment]
> TL;DR context, I'm suggesting that rather than overriding the _brew() completion function from the _brew_cask() we flip this so that _brew() calls _brew_cask() (or any other brew subcommand).

And TL;DR, yes, I think doing this makes sense as the basic approach to subcommand completion, including `cask`.

**Comment by joshka** [comment]
@pstadler thanks for your comment

> Minor doco thing: The joshka/brew-cask-zsh-completions repo should have its own standalone installation instructions which are independent of OMZ/Prezto/brew, even if that's as minimal as "copy this to your Zsh site-functions directory", and should reference the homebrew/completions tap and installation method in its README
Agreed.
>I think once this is working well, we should consider folding your completion definitions in to homebrew/brew itself, now that we are integrating Cask in to core Homebrew. It was Cask's policy to keep zsh completions as a separate project, but not core Homebrew's policy: core supplies them as part of the main Homebrew project. So if we migrate Cask to Homebrew core, it would be nice to have the completions for it available by default (and versioned to match the commands they go with).
I agree. In the mean time however I'd like things to work with the least manual steps possible. I think the separate repo will be ok until cask is part of homebrew core.

**Comment by joshka** [comment]
@apjanke I updated the doco https://github.com/joshka/brew-cask-zsh-completions/blob/master/README.md

**Comment by apjanke** [comment]
> I agree. In the mean time however I'd like things to work with the least manual steps possible. I think the separate repo will be ok until cask is part of homebrew core.

Yeah, no rush on this. And having the separate repo will allow you to iterate faster if changes to the completion definitions are needed.

**Comment by MikeMcQuaid** [comment]
@apjanke Any objections to 🚢 ing this as-is?

**Comment by apjanke** [comment]
No objections, as long as we're not considering this part of the "Homebrew public API" and committing to keeping the mechanism exactly as is. (And my understanding is that we aren't here, so :+1:.)

**Comment by joshka** [comment]
Any chance someone might merge this so the other parts can proceed?

**Comment by MikeMcQuaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by joshka** [comment]
Thanks @MikeMcQuaid

**Comment by joshka** [comment]
Hey - I'm looking at how the cask command is now part of homebrew directly. This change happened between initially opening this PR and now. As such, I think it now makes a bunch of sense to include the brew_cask completion directly in this repo rather than a sub-repo. I'll open a new PR for this shortly.
